stdlib/asm/crypto/fri/frif4e2.masm

```diff
@@ -0,0 +1,247 @@
+export.verify_remainder_query               
+
+    # Compute the index 
+    movup.2
+    push.2
+    u32checked_divmod   #[f_pos % 2, ?, ptr2n, (g^p)^2^n, ne1, ne0, ptr2n, ..]
+    movup.2
+    add
+    push.0.0.0.0
+    movup.4
+    mem_loadw           #[v1, v0, v1', v0', ?, (g^p)^2^n, ne1, ne0, ptr2n, ..]
+    movup.2
+    swap
+    dup.4
+    cdrop
+    swap.3
+    cdrop               #[v1, v0, x, ne1, ne0, ptr2n]
+    movup.3
+    assert_eq
+    swap drop
+    assert_eq
+    drop
+end
+
+#! Checks that, for a query with index p at layer i, the folding procedure to create layer (i + 1)
+#! was performed correctly.      
+#! Input: [ptr0 + 2 * i, g^p, p, e1, e0, ptr2n, ..]
+#! Output: [ptr0 + 2 * (i + 1), (g^p)^4, f_pos, ne1, ne0, ptr2n, ..]
+#! Cost: 98
+export.verify_query_layer.2
+
+    
+    # Fetch C and [d_size, t_depth, a1, a0] from memory
+    # [C, d_size, t_depth, a1, a0, g^p, p, e1, e0, ptr0, ptr2n, ..]
+    # 18 cycles
+    movdn.4
+    dup.4
+    push.0.0.0.0
+    dup.4
+    mem_loadw
+    movup.4
+    push.0.0.0.0
+    movup.4
+    mem_loadw
+
+    
+    # Verify Merkle auth path for (index = p % d_size, depth = t_depth, Root = C)
+    # [V, p % d_size, d_seg, g^p, e1, e0, a1, a0, ptr0, ptr2n, ..]
+    # 27 cycles
+    swapw.2                 # [g^p, p, e1, e0, d_size, t_depth, a1, a0, C, ptr0, ptr2n, ..]
+    swap                    # [p, g^p, e1, e0, d_size, t_depth, a1, a0, C, ptr0, ptr2n, ..]
+    movup.4                 # [d_size, p, g^p, e1, e0, t_depth, a1, a0, C, ptr0, ptr2n, ..]
+    u32checked_divmod       # [p % d_size, d_seg, g^p, e1, e0, t_depth, a1, a0, C, ptr0, ptr2n, ..]
+    movup.5                 # [t_depth, p % d_size, d_seg, g^p, e1, e0, a1, a0, C, ptr0, ptr2n, ..]
+    movupw.2                # [C, t_depth, p % d_size, d_seg, g^p, e1, e0, a1, a0, ptr0, ptr2n, ..]
+    dup.5
+    movup.5                 # [t_depth, p % d_size, C, p % d_size, d_seg, g^p, e1, e0, a1, a0, ptr0, ptr2n, ..]
+    mtree_get               # [t_depth, p % d_size, C, p % d_size, d_seg, g^p, e1, e0, a1, a0, ptr0, ptr2n, ..]
+    swapw dropw             # [V, p % d_size, d_seg, g^p, e1, e0, a1, a0, ptr0, ptr2n, ..]
+
+    # Unhashing the leaf values i.e. V to (v7,..., v0)
+    # Output: [v7, ..., v0, ]
+    # Cost: ~52 cycles
+
+    # Unhash V and save the pre-image in locaddr.0 and locaddr.1
+    # [T2, T1, T0, V, p % d_size, d_seg, g^p, e1, e0, a1, a0, ptr0, ptr2n, ..]
+    # 17 cycles
+    locaddr.1
+    push.0.0.0.0.0.0.0.0.0.0.0.0
+    adv_pipe
+    adv_pipe
+
+    # Assert T1 == V
+    # 13 cycles
+    swapw.3
+    movup.4
+    assert_eq
+    movup.3
+    assert_eq
+    movup.2
+    assert_eq
+    assert_eq 
+    
+    # Load (v7, ..v0) from memory
+    # [v7, ..., v0, p % d_size, d_seg, g^p, e1, e0, a1, a0, ptr0, ptr2n, ..]
+    # 3 cycles
+    loc_loadw.2
+    swapw
+    loc_loadw.1
+
+    # Call exec.FRIFOLD4
+    # [x, x, x, x, x, x, x, x, x, x, x, (g^p)^4, p % d_size, ne1, ne0, ptr0 + 2, ptr2n, ..] 
+    # 1 cycle
```

Comment:
I'm wondering if the output here could look as follows:
```
[ptr0+2, x, x, x, x, x, x, x, ptr0+2, ptr0+3, ptr0+2, (g^p)^4, p % d_size, ne1, ne0, ptr2n, ...]
```
This way, the end of procedure could be replaced by:
```
swapdw
dup.7
neq
```
And this would take 4 cycles instead of 18.

This also means that we'd enter the next iteration of the loop in the following state:
```
[ptr0+3, ptr0+2, (g^p)^4, p % d_size, ne1, ne0, ptr2n, ptr0+2, x, x, x, x, x, x, x, ...]
```
And then we can load values from memory like this:
```
pad
swapdw
mem_loadw
swapw
movup.8
mem_loadw
```
This would be 8 cycles instead of current 18.

We will need to make sure the stack is arranged correctly for the first iteration of the loop, but I think the net savings should be positive (we should be saving 26 cycles per layer).

Reply:
Cool! I didn't know we had `swapdw`. I will update the code with your suggestion.

Reply:
Sounds good! I will work from now onward with this layout in mind.

Reply:
I think I got a bit too aggressive with duplicating values in the output - we actually need most of these `x`'s for constraints. So, here is an updated scheme:
```
Input: [v7, v6, v5, v4, v3, v2, v1, v0, p % d_size, d_seg, g^p, e1, e0, a1, a0, ptr0, ptr2n, ..]
Output: [x, x, x, x, x, x, x, x, x, x, ptr0+2, (g^p)^4, p % d_size, ne1, ne0, ptr2n, ...]
```
With this scheme, the end of the procedure could look like:
```
swapdw
dup.7
dup.3
neq
```
And then the beginning of the loop could look like:
```
swapdw
drop
dup.9
add.1
mem_loadw
swapw
dup.9
mem_loadw
movup.8
drop
```
This is not as efficient as the previous approach, but we still should be saving 20 cycles per iteration.

Reply:
I don't see how `ptr0 + 2 * i` is kept on the stack with the above stack manipulations.
I am getting that the loop exists with the stack in the following configuration:
`[?, x, x, ptr0+2, (g^p)^4, p % d_size, ne1, ne0, ptr2n, x, x, x, x, x, x, x, x, ...]`

But:
```
swapdw
drop
dup.9
add.1
mem_loadw
swapw
dup.9
mem_loadw
movup.8
drop
```
leaves the stack as:
`[Word0, Word1, (g^p)^4, p % d_size, ne1, ne0, ptr2n,  ...]`

Reply:
I think it would go like this:
```
# start: [x, x, ptr0+2, (g^p)^4, p % d_size, ne1, ne0, ptr2n, x, x, x, x, x, x, x, x, ...]

swapdw     # [x, x, x, x, x, x, x, x, x, x, ptr0+2, (g^p)^4, p % d_size, ne1, ne0, ptr2n, ...]
drop       # [x, x, x, x, x, x, x, x, x, ptr0+2, (g^p)^4, p % d_size, ne1, ne0, ptr2n, ...]
dup.9      # [ptr0+2, x, x, x, x, x, x, x, x, x, ptr0+2, (g^p)^4, p % d_size, ne1, ne0, ptr2n, ...]
add.1      # [ptr0+3, x, x, x, x, x, x, x, x, x, ptr0+2, (g^p)^4, p % d_size, ne1, ne0, ptr2n, ...]
mem_loadw  # [Word1, x, x, x, x, x, ptr0+2, (g^p)^4, p % d_size, ne1, ne0, ptr2n, ...]
swapw      # [x, x, x, x, Word1, x, ptr0+2, (g^p)^4, p % d_size, ne1, ne0, ptr2n, ...]
dup.9      # [ptr0+2, x, x, x, x, Word1, x, ptr0+2, (g^p)^4, p % d_size, ne1, ne0, ptr2n, ...]
mem_loadw  # [Word0, Word1, x, ptr0+2, (g^p)^4, p % d_size, ne1, ne0, ptr2n, ...]
movup.8    # [x, Word0, Word1, ptr0+2, (g^p)^4, p % d_size, ne1, ne0, ptr2n, ...]
drop       # [Word0, Word1, ptr0+2, (g^p)^4, p % d_size, ne1, ne0, ptr2n, ...]
```
But maybe I missed something?

Reply:
Makes sense, I miscounted one of the garbage stack elements. Thank you!